src/D2L.CodeStyle.Analyzers/Async/Generator/SourceFilePathHintNameGenerator.cs

```diff
@@ -0,0 +1,110 @@
+﻿#nullable enable
+
+using System.Collections.Immutable;
+using System.Text.RegularExpressions;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator {
+	/// <summary>
+	/// Given a stream of input source paths generate a nice looking set of
+	/// distinct hint names for the source generator.
+	/// </summary>
+	internal sealed class SourceFilePathHintNameGenerator {
+		private readonly ImmutableArray<string> m_sourcePaths;
+		public readonly int m_prefixLength;
+
+		private SourceFilePathHintNameGenerator(
+			ImmutableArray<string> sourcePaths,
+			int prefixLength
+		) {
+			m_sourcePaths = sourcePaths;
+			m_prefixLength = prefixLength;
+		}
+
+		public static SourceFilePathHintNameGenerator Create( ImmutableArray<string> sourcePaths ) {
+			// Calculate the longest prefix of all input strings
+
+			if( sourcePaths.Length == 0 ) {
+				return new SourceFilePathHintNameGenerator( sourcePaths, 0 );
+			}
+
+			// Seed it with the first source files directory
+			string commonPrefix = RemoveFileName( sourcePaths[0] );
+
+			foreach( var sourcePath in sourcePaths.Select( RemoveFileName ) ) {
+				int idx = 0;
+				int maxPrefixLength = Math.Min( sourcePath.Length, commonPrefix.Length );
+
+				while( idx < maxPrefixLength ) {
+					if( sourcePath[idx] != commonPrefix[idx] ) {
+						break;
+					}
+
+					idx++;
+				}
+
+				if( idx != commonPrefix.Length ) {
+					commonPrefix = sourcePath.Substring( 0, idx );

```

Comment:
This could lead to a possible noisy diff if we check these files in.  Like all of a sudden someone checks in on new file, and 50 others will show as changes because a new common path is established

Reply:
Yep.